### ISSUE TITLE
Derived type arguments are dispatched to hub methods as null

### ISSUE THREAD

**Issue by NickDarvey** [aspnet/AspNetCore#11495]
...but only when you include a synthetic parameter in your hub method's signature.

### Describe the bug
I'm invoking a hub method and sending a tree where each node might be a subclass of some root node class. The deserializer correctly deserializes each node into their subclass so the runtime type of the argument is not the exact type as in the method signature, but rather a derived one.

This works correctly _unless_ a synthetic parameter is included in the method, i.e.  a CancellationToken token parameter. If you include a CancellationToken, the derived arguments will be dispatched as nulls to the hub method.

### To Reproduce
1. Create a class relationship like:
   ```csharp
   public class ANode {}
   public class BNode : ANode {}
   ```
2. And a hub method like:
   ```csharp
   public async IAsyncEnumerable<string> DoThing(ANode param, [EnumeratorCancellation]CancellationToken token) { /* ... */ }
   ```
3. Invoke the hub method with a `BNode` (e.g. by using Json.NET's `TypeNameHandling = TypeNameHandling.All`)

### Expected behavior
Hub method receives a `BNode` as `param`.

### Actual behavior
Hub method receives `null` as `param`.

### Cause
When there are [synthetic arguments](https://github.com/aspnet/AspNetCore/blob/5b31a9540aa64dcf613ce423c25eb630b84b0fac/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L254), the original arguments from the `HubInvocationMesssage` are looped through and matched against the `HubMethodDescriptor`'s parameters to rebuild the arguments with the added synthetic parameters. [It uses equality comparison to see if the argument fits the parameter](https://github.com/aspnet/AspNetCore/blob/5b31a9540aa64dcf613ce423c25eb630b84b0fac/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L264), which won't match on derived types. In scenario above [the code path labelled 'This should never happen'](https://github.com/aspnet/AspNetCore/blob/5b31a9540aa64dcf613ce423c25eb630b84b0fac/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L288) does happen and the current index of the arguments is left null.

### Additional context
ASP.NET Core 3.0.0-preview6.19307.2

**Comment by anurse** [comment]
Ooh, nice catch, and thanks for the corresponding PR.

I should caution you that usage of `TypeNameHandling.All` is a *serious security risk* since a malicious client could specify absolutely any type name and the JSON serializer would construct it, thus running any of the code in it's constructor. This is a remote code execution risk. A much better solution is to implement a custom `type` field with a **fixed set of known values** and write a [custom `JsonConverter`](https://www.newtonsoft.com/json/help/html/CustomJsonConverter.htm) that can look up the matching type and instantiate it.

If you have to use `TypeNameHandling.All`, you should **always** use a custom [`SerializationBinder`](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializer_SerializationBinder.htm) when using an option like this. Implement `BindToType` and **only** allow types you recognize and **know to be safe** to be bound. In this case, the list of known sub-types should be used.

You should use these protections **even when the client code is trusted** for defense in depth.

**Comment by NickDarvey** [comment]
Thanks for highlighting the security issues, a `KnownTypesBinder` is what I have at the moment. Though, I was looking at switching to System.Text.Json so I think I might end up with the converter approach you mentioned.

**Comment by anurse** [comment]
> a `KnownTypesBinder` is what I have at the moment.

👍 Glad to hear that, I just wanted to get on my soapbox and make sure you (and anyone who comes across the issue) were aware of the potential security risk :).

Even with a `KnownTypesBinder` I'd definitely suggest looking at the `JsonConverter` model. It decouples your message format from the specific type names (so you could refactor and simply adjust your converter) and makes it absolutely certain that the client can't activate an unexpected type on the server. Also, as you noted, it will be more compatible with `System.Text.Json`, which is adding a `JsonConverter` pattern (very similar in API to Json.NET).

**Comment by halter73** [comment]
Fixed by #11496
